### PR TITLE
Add Migration for Cost Sources to use Application Authentication resources

### PR DESCRIPTION
**Pull request by lindgrenj6**
This will make the availability check always fall back to the Application status, since no endpoint is available.

Tested locally and the fallback code works :+1:, now we just need the UI to create them under the application instead. This should probably go in after that UI change so we don't have to run this again.

JIRA: https://projects.engineering.redhat.com/browse/TPINVTRY-1082

\# TODO:
- [x] Cost Management updated to use AppAuth vs EndpointAuth
- [x] Subscription Watch updated to use AppAuth vs EndpointAuth
- [x] CM UI Updated to create AppAuth instead of EndpointAuth and update those as well
- [x] SW UI Updated to create AppAuth instead of EndpointAuth and update those as well

cc @syncrou @dccurtis

### CODE REVIEW

**Comment by lindgrenj6** [comment]
cc @rvsia

**Comment by syncrou** [comment]
@lindgrenj6 was it expected that the specs were red?

**Comment by dccurtis** [comment]
@lindgrenj6 Can you hold off on merging this.  It looks like there are some small changes I need to make on the Cost side.

**Comment by lindgrenj6** [comment]
@dccurtis yep, we probably aren't going to merge this until the UI is fixed :+1:

**Comment by dccurtis** [comment]
@lindgrenj6 https://github.com/project-koku/koku/pull/2398 has been merged.  Cost supports both Application and Endpoint authentications with this change.

**Comment by rvsia** [comment]
@lindgrenj6 The ui is ready (https://github.com/RedHatInsights/sources-ui/pull/438) - for cost sources (aws/azure) it does not create empty endpoints anymore.

File: db/migrate/20200911162327_move_endpoint_authentications_to_application_authentication.rb
Comment (slemrmartin) [inline db/migrate/20200911162327_move_endpoint_authentications_to_application_authentication.rb:10]: @lindgrenj6 
We have also `aplication_authentications` table
model `Application`: `has_many :authentications, :through => :application_authentications`

We should review if there is use case for this relation and remove it, if not, it's confusing

Comment (slemrmartin) [reply db/migrate/20200911162327_move_endpoint_authentications_to_application_authentication.rb:10]: @rvsia is UI using this M:N relationship somewhere?

Comment (rvsia) [reply db/migrate/20200911162327_move_endpoint_authentications_to_application_authentication.rb:10]: @slemrmartin Yeah. It connects all applications with their assigned authentications.

Comment (slemrmartin) [reply db/migrate/20200911162327_move_endpoint_authentications_to_application_authentication.rb:10]: ok, it means this migration should use the applications_authentications too

Comment (rvsia) [reply db/migrate/20200911162327_move_endpoint_authentications_to_application_authentication.rb:10]: @slemrmartin I think these relationships should already exist. At least, the UI creates them since the table was introduced.

However, with the latest changes, we have got two 'connections' between an app and auths.

I am kinda discussing this issue here https://issues.redhat.com/browse/RHCLOUD-9526

**Comment by werwty** [comment]
The Subscription Watch changes to use AppAuth have been merged.

**Comment by lindgrenj6** [comment]
@rvsia @dccurtis @syncrou 

So it sounds like this is about ready to be merged - @rvsia I remember hearing that we can just do a :heavy_check_mark: in the config for SW UI to turn on the changes? If thats true then we're probably good to go here.

**Comment by rvsia** [comment]
@lindgrenj6 Yeah, working on that now.

**Comment by rvsia** [comment]
@lindgrenj6 The UI is ready! :crossed_swords:

**Review by syncrou — APPROVED** [review]
It seems like this is ready to go, but there are no Approvals yet.  

LGTM 👍 

@slemrmartin is this ready to go?

**Comment by slemrmartin** [comment]
@syncrou @lindgrenj6 no, we should discuss my review as this is breaking intended architecture.

**Comment by slemrmartin** [comment]
and read my comment https://issues.redhat.com/browse/RHCLOUD-9526 please, it's related to this and the future

**Comment by lindgrenj6** [comment]
After speaking with @slemrmartin to better understand the problem I was unaware that `ApplicationAuthentication` was an actual table - I wasn't thinking that was a table due to the fact we have the polymorphic relationship on the Authentication model. 

I added some logic to dynamically create them on up, remove them on down.

**Review by slemrmartin — APPROVED** [review]
:+1: LGTM

**Comment by lindgrenj6** [comment]
@roliveri @syncrou I think this is good to go.

**Pull request merged**
